mypy/plugins/singledispatch.py

```diff
@@ -0,0 +1,178 @@
+    """Get the element that corresponds to the first argument passed to the function"""
+    if args and args[0]:
+        return args[0][0]
+    return None
+
+
+REGISTER_RETURN_CLASS = 'SingleDispatchRegisterCallable'
+
+
+def make_fake_register_class_instance(api: CheckerPluginInterface) -> Instance:
+    defn = ClassDef(REGISTER_RETURN_CLASS, Block([]))
+    defn.fullname = 'functools.{}'.format(REGISTER_RETURN_CLASS)
+    info = TypeInfo(SymbolTable(), defn, "functools")
+    obj_type = api.named_generic_type('builtins.object', []).type
+    info.bases = [Instance(obj_type, [])]
+    info.mro = [info, obj_type]
+    defn.info = info
+
+    function_type = api.named_generic_type('builtins.function', [])
+
+    func_arg = Argument(Var('name'), AnyType(TypeOfAny.implementation_artifact), None, ARG_POS)
+    add_method_to_class_with_function_type(function_type, defn, '__call__', [func_arg], NoneType())
+
+    return Instance(info, [])
+
+
+def create_singledispatch_function_callback(ctx: FunctionContext) -> Type:
+    # TODO: check that there's only one argument
+    func_type = get_proper_type(get_first_arg(ctx.arg_types))
+    if isinstance(func_type, CallableType):
+        # TODO: support using type as argument to register
+        metadata = {
+            'fallback': func_type,
+            'registered': {}
+        }  # type: SingledispatchInfo
+
+        # singledispatch returns an instance of functools._SingleDispatchCallable according to
+        # typeshed
+        singledispatch_obj = get_proper_type(ctx.default_return_type)
+        assert isinstance(singledispatch_obj, Instance)
+        # mypy shows an error when assigning TypedDict to a regular dict
+        singledispatch_obj.type.metadata[METADATA_KEY] = metadata  # type: ignore
+
+    return ctx.default_return_type
+
+
+def singledispatch_register_callback(ctx: MethodContext) -> Type:
+    if isinstance(ctx.type, Instance):
+        # TODO: check that there's only one argument
+        first_arg_type = get_proper_type(get_first_arg(ctx.arg_types))
+        if isinstance(first_arg_type, (CallableType, Overloaded)) and first_arg_type.is_type_obj():
+            # HACK: We receieved a class as an argument to register. We need to be able
+            # to access the function that register is being applied to, and the typeshed definition
+            # of register has it return a generic Callable, so we create a new
+            # SingleDispatchRegisterCallable class, define a __call__ method, and then add a
+            # plugin hook for that.
+
+            # is_subtype doesn't work when the right type is Overloaded, so we need the
+            # actual type
+            register_type = first_arg_type.items()[0].ret_type
+            register_callable = make_fake_register_class_instance(ctx.api)
+            register_metadata = {
+                'register_type': register_type,
+                'singledispatch_obj': ctx.type
+            }  # type: RegisterCallableInfo
```

Comment:
I think I'd prefer storing the types as type arguments to `_SingleDispatchRegisterCallable`. (In the `args` field of `Instance`). So I guess make `make_fake_register_class_instance` return the `TypeInfo` instead and then construct the Instance here?

Reply:
This should be fixed in 61172ce823f21825d01199ed5b8de3137d1282ca.